src/filter.ts

```diff
@@ -524,15 +532,40 @@ type ExpandL2<S extends string> = S extends L2Shortcuts
     : S;
 
 // === Define some helpers for when one property implies the existence of others
-type Twins<V extends string> = V extends KeyOf<Equivalents> ? Equivalents[V]
-    : V;
-type Equivalents = {
+
+// merges twins based on L1 with those based on L1 and L2
+type AddTwins<L1 extends string, L2 extends string> =
+    | TwinsFromL1<L1, L2>
+    | TwinsFromL2<L1, L2>;
+
+// yields twins based on a given L1 property
+type TwinsFromL1<L1 extends string, L2 extends string> = L1 extends
+    KeyOf<L1Equivalents> ? L1Equivalents[L1]
+    : L2;
+type L1Equivalents = {
+    message: "from";
+    edited_message: "from" | "edit_date";
+    channel_post: "sender_chat";
+    edited_channel_post: "sender_chat" | "edit_date";
+};
+
+// yields twins based on given L1 and L2 properties
+type TwinsFromL2<L1 extends string, L2 extends string> = L1 extends
+    KeyOf<L2Equivalents>
+    ? L2 extends KeyOf<L2Equivalents[L1]> ? L2Equivalents[L1][L2] : L2
+    : L2;
```

Comment:
Old `Twins` implementation, but now takes into account that twins (same-level) are only applied behind the right L1 properties.

For example, before we would also set `document` when given `animation` inside `callback_query` updates, and the only reason this was not a problem is because those clashes coincidentally do not appear.

Now, we make sure that twins are only applied behind the correct update types. This is also a lot more flexible for future API changes.